### PR TITLE
Account for remaining recovery in disk allocator

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the disk-based shard allocator accounts for incoming shards by
subtracting the estimated size of the incoming shard from the free space on the
node. This is an overly conservative estimate if the incoming shard has almost
finished its recovery since in that case it is already consuming most of the
disk space it needs.

This change adds to the shard stats a measure of how much larger each store is
expected to grow, computed from the ongoing recovery, and uses this to account
for the disk usage of incoming shards more accurately.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Although I ran the `:server` test suite a good deal, CI is indicating that there's quite a few more test adjustments needed. Suggest holding off on detailed reviews until I've made CI happy; marking this as WIP for now.

**Comment by DaveCTurner** [comment]
Ok it was a lot of noise from relatively few broken tests, this is good to go now.

**Review by ywelsch — COMMENTED** [review]
I've left some minor comments, overall looking good though.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java [hunk 1]
```diff
@@ -383,16 +395,28 @@ public void addListener(Consumer<ClusterInfo> clusterInfoConsumer) {
...
+
+            final String shardIdentifier = ClusterInfo.shardIdentifierFromRouting(shardRouting);
             if (logger.isTraceEnabled()) {
-                logger.trace("shard: {} size: {}", sid, size);
+                logger.trace("shard: {} size: {} reserved: {}", shardIdentifier, size, reserved);
+            }
+            shardSizes.put(shardIdentifier, size);
+
+            if (reserved != StoreStats.UNKNOWN_RESERVED_BYTES) {
+                reservedSpaceByShard.computeIfAbsent(
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:414]: I would have thought this should merge the contents to accommodate for all shards on a specific NodePath rather than only the first encountered? Or am I missing something?

File: server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java [hunk 2]
```diff
@@ -109,6 +127,14 @@ public void writeTo(StreamOutput out) throws IOException {
             c.key.writeTo(out);
             out.writeString(c.value);
         }
+
+        if (out.getVersion().onOrAfter(StoreStats.RESERVED_BYTES_VERSION)) {
+            out.writeVInt(this.reservedSpace.size());
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java:132]: An `ImmutableOpenMap` isn't a `Map` 😢

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -1004,8 +1004,24 @@ public GetStats getStats() {
     }
 
     public StoreStats storeStats() {
+        final LongSupplier remainingRecoveryBytesSupplier;
+        synchronized (mutex) {
+            if (recoveryState == null) {
+                remainingRecoveryBytesSupplier = () -> 0L;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Ah you're right, this indicates that recovery hasn't started, not that no recovery is required. I pushed 6a7060c3d7d.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -1004,8 +1004,24 @@ public GetStats getStats() {
     }
 
     public StoreStats storeStats() {
+        final LongSupplier remainingRecoveryBytesSupplier;
+        synchronized (mutex) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:1008]: `recoveryState` is nonvolatile and assigned under `mutex` so I think we cannot avoid synchronising on `mutex` here. Note that we do almost no work under the mutex itself. Would you rather make `recoveryState` volatile?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java [hunk 5]
```diff
@@ -140,6 +140,10 @@ public void onNewInfo(ClusterInfo info) {
             final DiskUsage usage = entry.value;
             final RoutingNode routingNode = routingNodes.node(node);
 
+            final long reservedSpace = info.getReservedSpace(usage.getNodeId(), usage.getPath()).getTotal();
+            final DiskUsage usageWithReservedSpace = new DiskUsage(usage.getNodeId(), usage.getNodeName(), usage.getPath(),
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:144]: AFAICS this compensation affects two functionalities:
1. Triggering a reroute when going under low threshold
2. Auto-releasing indices when disk usage goes below high threshold.

I think 1) is fine, but I am a bit concerned about 2), since the primary reason we only release when under high threshold is to have some hysteresis to not trigger flood-stage on/off rapidly. I think I would prefer to do the auto-release based on the non-compensated number (just like we move to flood-stage based on that), but am curious on your thoughts on this.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:1008]: yeah, I'd prefer that. I'm scared of stats calls going under this mutex, as we currently do some heavy operations under that mutex (e.g. flush on closing), which ofc we should not.

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Ok I pushed a5275ed96d0

File: server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java [hunk 6]
```diff
@@ -403,13 +404,15 @@ private void internalRecoverFromStore(IndexShard indexShard) throws IndexShardRe
...
                 // since we recover from local, just fill the files and size
+                final RecoveryState.Index index = recoveryState.getIndex();
                 try {
-                    final RecoveryState.Index index = recoveryState.getIndex();
                     if (si != null) {
                         addRecoveredFileDetails(si, store, index);
                     }
                 } catch (IOException e) {
                     logger.debug("failed to list file details", e);
+                } finally {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java:414]: Not sure I understand why this needs to be in a finally?

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java [hunk 7]
```diff
@@ -98,9 +99,16 @@ public DiskThresholdDecider(Settings settings, ClusterSettings clusterSettings)
      */
     public static long sizeOfRelocatingShards(RoutingNode node, boolean subtractShardsMovingAway, String dataPath, ClusterInfo clusterInfo,
                                               Metadata metadata, RoutingTable routingTable) {
-        long totalSize = 0L;
-
-        for (ShardRouting routing : node.shardsWithState(ShardRoutingState.INITIALIZING)) {
+        // Account for reserved space wherever it is available
+        final ClusterInfo.ReservedSpace reservedSpace = clusterInfo.getReservedSpace(node.nodeId(), dataPath);
+        long totalSize = reservedSpace.getTotal();
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:104]: We count reserved space as only the space for the original files in phase 1 of recovery. This makes us slightly more optimistic than desired, in that any changes coming in after getting the original file list are not accounted for as reserved. For instance in cases like described in #58011.

I wonder if it (for recoveries and relocations) was more precise to adjust by adding (primarySize - alreadyRecovered) - or just (primarySize - replicaSize), where we track each replica individually? This unfortunately does not give us snapshot, which the presented mechanism does, i.e. , we probably need both.

**Review by henningandersen — COMMENTED** [review]
I did an initial review and left some initial comments.

File: docs/reference/cluster/stats.asciidoc [hunk 8]
```diff
@@ -231,6 +231,18 @@ Total size of all shards assigned to selected nodes.
 `size_in_bytes`::
 (integer)
 Total size, in bytes, of all shards assigned to selected nodes.
+
+`reserved`::
+(<<byte-units,byte value>>)
+A prediction of how much larger the shard stores on this node will eventually
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc]: Thanks, yes. Actually not quite cluster-wide since you can select a subset of nodes here but "these nodes" is awkward so I removed it in https://github.com/elastic/elasticsearch/pull/58029/commits/b89a56770caa137261dd7f5e03336b7b22af2012.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java [hunk 9]
```diff
@@ -318,15 +323,21 @@ public void onFailure(Exception e) {
...
+                final ImmutableOpenMap.Builder<ShardRouting, String> dataPathByShardRoutingBuilder = ImmutableOpenMap.builder();
+                final Map<ClusterInfo.NodeAndPath, ClusterInfo.ReservedSpace.Builder> reservedSpaceBuilders = new HashMap<>();
+                buildShardLevelInfo(logger, stats, shardSizeByIdentifierBuilder, dataPathByShardRoutingBuilder, reservedSpaceBuilders);
+
+                final ImmutableOpenMap.Builder<ClusterInfo.NodeAndPath, ClusterInfo.ReservedSpace> rsrvdSpace = ImmutableOpenMap.builder();
+                reservedSpaceBuilders.forEach((nodeAndPath, builder) -> rsrvdSpace.put(nodeAndPath, builder.build()));
+
+                shardSizes = shardSizeByIdentifierBuilder.build();
+                shardRoutingToDataPath = dataPathByShardRoutingBuilder.build();
+                reservedSpace = rsrvdSpace.build();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java]: Race conditions abound here. We're also relying on the disk usage from the nodes stats which may be received at a completely different time, we don't synchronise the collection of stats on the nodes themselves either, and the store size comes from a cached value anyway. I adjusted this bit of code in https://github.com/elastic/elasticsearch/pull/58029/commits/9639659aa690679fa4f181885724a7dfd35b32e6 but that's certainly not a complete fix.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:411]: > am I missing something?

Possibly a closing bracket :grin: We compute a new _builder_ if absent, but then `.add()` the stats to the builder for each relevant shard.

File: server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java (hunk 6 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java:414]: IIRC an `IOException` here can happen if force-allocating a stale primary, but is not fatal. I didn't investigate further since it doesn't really matter for our purposes: we reuse all the local files anyway so the reserved space is zero.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:144]: Yes, this is a good point. I'll work on addressing that.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java (hunk 7 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDecider.java:104]: We discussed this via another channel. The idea to use (primarySize - replicaSize) does not work well in the partial recovery case, since we do not remove existing files on the target until the end of phase 1.

I think the replacement here is good enough.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:144]: I'm no longer sure about this. The compensated disk usage is never less than the non-compensated one, so auto-releasing based on the compensated number has _more_ hysteresis than today's behaviour. Is the additional hysteresis the thing that concerns you @henningandersen or has one of us got our logic backwards?

File: server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java (hunk 6 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java:414]: Still not sure why the finally block is necessary, since the catch clause for IOException does not rethrow?

It seems illogical that in this block a finally block is necessary, whereas in the next block it is not?

I mean it will work fine either way... So feel free to disregard this if you prefer.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 5 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:144]: Yes, that is my concern. Something cured the situation (deleted index, moved shards etc.), but due to an otherwise unrelated recovery, we would keep halting indexing until that recovery is done. I would prefer to resume indexing (i.e., release the block) and then perhaps later find out that we need to halt it again. If cluster is truly full, we will get to that point, but if it is has enough space, we are likely to cure it before hitting the flood stage again anyway.

I think the additional hysteresis is unnecessary. AFAIK, we have not seen that the block was flapping too frequently and using the non-compensated number, we would simply keep the behavior unmodified from this PR.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:411]: I see it now ! Thanks.

nit: maybe put the `.add` part on the next line.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java]: Agreed and thanks for the change, looks good.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java]: Sure, I introduced an intermediate variable in bf9faca0d0d6a8859c821f165847267f66153d6a.

File: server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java (hunk 6 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/index/shard/StoreRecovery.java]: Oh, I see now. Fixed in cf23d43f59fb1330fd0ab791e3f772654b6825f1.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:144]: I see, right. I don't think the additional hysteresis would be much of a problem in practice (if you hit the flood-stage watermark you already lost) and continuing to recover onto a node that's breached the flood-stage watermark is likely the real bug here. However it turns out not to add too much complexity so I adjusted this in d8afaf36f6d723a485c39dd53c26a3d3dbdcf7cf.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java [hunk 10]
```diff
@@ -200,7 +201,9 @@ public void clusterChanged(ClusterChangedEvent event) {
 
     @Override
     public ClusterInfo getClusterInfo() {
-        return new ClusterInfo(leastAvailableSpaceUsages, mostAvailableSpaceUsages, shardSizes, shardRoutingToDataPath);
+        final IndicesStatsSummary indicesStatsSummary = this.indicesStatsSummary;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java:204]: add comment here // single volatile read

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 5 above)
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:144]: Can you move the creation of this object to the place where it's first used? Or should we  alternatively call the other `usage` instead `usageWithoutReservedSpace` to make it clearer  when we are using one vs the other

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java [hunk 11]
```diff
@@ -158,18 +162,26 @@ public void onNewInfo(ClusterInfo info) {
                 logger.warn("flood stage disk watermark [{}] exceeded on {}, all indices on this node will be marked read-only",
                     diskThresholdSettings.describeFloodStageThreshold(), usage);
 
-            } else if (usage.getFreeBytes() < diskThresholdSettings.getFreeBytesThresholdHigh().getBytes() ||
-                usage.getFreeDiskAsPercentage() < diskThresholdSettings.getFreeDiskThresholdHigh()) {
+                continue;
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:165]: Control flow is  a bit crazy in this method now. Any thoughts on simplifying it?

**Review by ywelsch — APPROVED** [review]
I've left a few nits, looking good o.w.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 5 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java]: I tried renaming it but I found that harder to distinguish (`usageWithoutReservedSpace` and `usageWithReservedSpace` are pretty similar to the eye). Moved the construction lower down in 56ecdb71c243a9fdac03826d8c5b210ccb18d23b.

File: server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java (hunk 10 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java]: :+1: 78d353e035146eb1f25c58ace0cff6548a3c83ca

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java (hunk 11 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:161]: Yes, I was considering introducing an explicit state machine for each node as I have always found this logic a little hard to follow. Not convinced it's worth it, I think the tests have pretty good coverage.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java:161]: ok, the tests are a good argument. Let's proceed as is then

**Review by henningandersen — COMMENTED** [review]
This looks good, I have just two comments/questions on exposing unknown size/-1 in the API.

File: docs/reference/cluster/stats.asciidoc [hunk 12]
```diff
@@ -231,6 +231,18 @@ Total size of all shards assigned to selected nodes.
 `size_in_bytes`::
 (integer)
 Total size, in bytes, of all shards assigned to selected nodes.
+
+`reserved`::
+(<<byte-units,byte value>>)
+A prediction of how much larger the shard stores will eventually grow due to
+ongoing peer recoveries, restoring snapshots, and similar activities. A value
+of `-1b` indicates that this is not available.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:239]: With this change the cluster-wide reserved size will indeed be zero as long as all nodes in the cluster are sufficiently new. It may be -1b in clusters with a mix of versions since older versions do not know how to supply this value.

With an upcoming change, however, it will be positive for some active shards (namely, searchable snapshots that are still warming up). It shouldn't be unknown for active shards on sufficiently new versions indeed, but I would still rather distinguish "nothing is reserved" from "unknown" in case this is helpful for some future analysis of a diagnostics bundle.

File: docs/reference/cluster/nodes-stats.asciidoc [hunk 13]
```diff
@@ -245,6 +245,18 @@ Total size of all shards assigned to the node.
 `size_in_bytes`::
 (integer)
 Total size, in bytes, of all shards assigned to the node.
+
+`reserved`::
+(<<byte-units,byte value>>)
+A prediction of how much larger the shard stores on this node will eventually
+grow due to ongoing peer recoveries, restoring snapshots, and similar
+activities. A value of `-1b` indicates that this is not available.
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/nodes-stats.asciidoc:253]: `GET _nodes/stats?level=shards` reports this value on a shard-by-shard basis, and includes initialising shards, so it could well be unknown in some of these cases.

File: docs/reference/cluster/stats.asciidoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply docs/reference/cluster/stats.asciidoc:239]: I think of that statement in the opposite way, in that I could be unlucky to get a diagnostics with -1 out and then know nothing about the reserved space at a node summary level.

Also, in the mixed cluster case, I would prefer to assume 0 from old versions. It would otherwise stay -1 at the summarized level until all data nodes are upgraded, which again hides the actual number.

File: docs/reference/cluster/nodes-stats.asciidoc (hunk 13 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply docs/reference/cluster/nodes-stats.asciidoc:253]: I think that on a shard by shard basis, -1 could make sense (though I find it hard to think of a case where it matters).

But on a summarized level, getting -1 out because the stats are pulled at an unfortunate time seems to hide more information than it gives.

File: docs/reference/cluster/stats.asciidoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply docs/reference/cluster/stats.asciidoc:239]: Ok, I pushed 6cee87b760a2338472b2272c1263cfdd2676888d with your idea to treat unknown as `0` in the aggregated stats, but still report -1 in the shard-level stats.

[the remark this answers is not in the document]
Comment (henningandersen) [reply docs/reference/cluster/stats.asciidoc:239]: I think we can then leave out the -1 notice here? May still need to be on the node-stats due to the shard level info, but AFAICS, it can never be -1 at cluster stats level?

**Review by henningandersen — APPROVED** [review]
LGTM.
